gluten-core/src/main/java/org/apache/gluten/memory/memtarget/spark/DynamicOffHeapSizingPolicyChecker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.memory.memtarget.spark;
+
+import org.apache.gluten.GlutenConfig;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public final class DynamicOffHeapSizingPolicyChecker {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(DynamicOffHeapSizingPolicyChecker.class);
+  private final long maxOnHeapMemoryInBytes = GlutenConfig.getConf().onHeapMemorySize();
+  private final AtomicLong usedOffHeapBytes = new AtomicLong();
+
+  DynamicOffHeapSizingPolicyChecker() {}
+
+  public boolean canBorrow(long size) {
+    if (size == 0) {
+      return true;
+    }
+
+    long totalMemory = Runtime.getRuntime().totalMemory();
```

Comment:
Wouldn't Runtime.getRuntime().maxMemory() - Runtime.getRuntime().totalMemory() be a better estimation for bytes used for on heap? 

Although (totalMemory - freeMemory) is the bytes used by _live_ objects, totalMemory is more representative of RSS memory. 

Any allocations by native would add on top of totalMemory.

Reply:
I believe this is the same point that zhztheplayer made above. Please see the reply to that thread.